src/lib/storage/table.cpp

```diff
@@ -104,17 +104,35 @@ const std::string& Table::column_name(ColumnID column_id) const { return _column
 
 const std::string& Table::column_type(ColumnID column_id) const { return _column_types.at(column_id); }
 
-Chunk& Table::get_chunk(ChunkID chunk_id) { return _chunks.at(chunk_id); }
+Chunk& Table::get_chunk(ChunkID chunk_id) { return *(_chunks.at(chunk_id)); }
 
-const Chunk& Table::get_chunk(ChunkID chunk_id) const { return _chunks.at(chunk_id); }
+const Chunk& Table::get_chunk(ChunkID chunk_id) const { return *(_chunks.at(chunk_id)); }
 
 bool Table::_chunk_matches_definitions() const {
   // Since we cannot alter column specifications after they have been created,
   // it suffices to check for the same length of columns in our definition and the
   // last chunk's definition
-  return _chunks.empty() || (_chunks.back().col_count() == _column_names.size());
+  return _chunks.empty() || (_chunks.back()->col_count() == _column_names.size());
 }
 
-void Table::compress_chunk(ChunkID chunk_id) { throw std::runtime_error("TODO"); }
+void Table::compress_chunk(ChunkID chunk_id) {
+  Assert(chunk_id < chunk_count(), "Index out of bounds");
+  Assert(chunk_id < chunk_count() - 1, "The last chunk must not be compressed");
+
+  const Chunk& uncompressed_chunk = get_chunk(chunk_id);
+  auto compressed_chunk = std::make_shared<Chunk>();
+
+  for (auto column_index = ColumnID{0}; column_index < uncompressed_chunk.col_count(); ++column_index) {
+    const auto& column_type = _column_types.at(column_index);
+    const auto& uncompressed_column = uncompressed_chunk.get_column(column_index);
+    auto compressed_column = make_shared_by_column_type<BaseColumn, DictionaryColumn>(column_type, uncompressed_column);
+    compressed_chunk->add_column(compressed_column);
+  }
+
+  _chunks.at(chunk_id) = compressed_chunk;
```

Comment:
Currently this works through an additional `std::shared_ptr` indirection, but will still work without it, as we can replace Chunks in-place through move semantics (as a Chunk is `NonCopyable`) (see comment in table.hpp)